### ISSUE TITLE
Checking COLA condition in istft should be optional (or omitted)

### ISSUE THREAD

**Issue by gauss256** [scipy/scipy#8791]
There is some confusion in the literature about the constant-overlap add (COLA) constraint and its relevance to the STFT and ISTFT. I have written [this blog post](https://gauss256.github.io/blog/cola.html) to clarify it. The gist of it is that (1) the COLA constraint is sufficient but not necessary for the STFT to be invertible, and (2) the appropriate COLA constraint for the STFT is on the _square_ of the window, not the window itself.

In SciPy, the `istft` function enforces COLA on the window (not its square) when `nperseg` and `noverlap` are specified. This is unnecessary, incorrect, and prevents the use of many useful windows. If you just comment out the `check_COLA` lines then `istft` perfectly inverts the STFT for all but the most pathological windows.

### Reproducing code example:
```python
import numpy as np
from scipy import signal

nperseg = 256
hop = 32
noverlap = nperseg - hop
x = np.random.randn(8 * 256)
w = signal.hann(nperseg)
w = np.sqrt(w)
f, t, Zxx = signal.stft(x, window=w, nperseg=nperseg, noverlap=noverlap)
_, y = signal.istft(Zxx, window=w, nperseg=nperseg, noverlap=noverlap)
print(f'allclose: {np.allclose(x, y)}')
```

### Error message:
With the `istft` code in its current form:
```
Traceback (most recent call last):
  File "eda_cola.py", line 11, in <module>
    _, y = signal.istft(Zxx, window=w, nperseg=nperseg, noverlap=noverlap)
  File "C:\Users\bruce\Anaconda3\lib\site-packages\scipy\signal\spectral.py", line 1053, in istft
    raise ValueError('Window, STFT shape and noverlap do not satisfy the '
ValueError: Window, STFT shape and noverlap do not satisfy the COLA constraint.
```
If the `check_COLA` lines are commented out, the output is 
```
allclose: True
```
The example is for root Hann, but almost any not-crazy combination of `window`, `nperseg`, and `noverlap` will work.

### Scipy/Numpy/Python version information:
```
1.0.1 1.14.2 sys.version_info(major=3, minor=6, micro=3, releaselevel='final', serial=0)
```

**Comment by larsoner** [comment]
@e-q you have mostly worked on the COLA code, do you have time to look into this? It does sound like a bug to me.

**Comment by e-q** [comment]
Interesting! I appreciate the time you've taken to look into this. I haven't gone through the math in your blog post in much detail, but there seem to be two actionable items here:
- We could relax the `check_COLA` into a function that checks that all window values are nonzero and that `noverlap>=0`.
- We allow for bypassing of any window checking in `istft`, perhaps through a new keyword argument.

If you have the time, it would be really helpful to see a pull request that implements these changes. In particular, it would be useful to see additions of windows that satisfy these relaxed conditions but not COLA in the "roundtrip" tests for `istft`:
https://github.com/scipy/scipy/blob/61efbf03d23fd3c4117f210becf6153c1cd6ddd9/scipy/signal/tests/test_spectral.py#L1063

**Comment by gauss256** [comment]
As you've noted, a related issue is that can be OK to have `noverlap==0`  but the code enforces `noverlap>0`.

Since use of COLA seems widespread (not sure why though!), I'd be inclined to leave `check_COLA` alone and make a new function `check _invertible` or something like that.

I'll think about it some more and see if I can make a pull request in the next day or two.

**Comment by mpariente** [comment]
I do agree on one very important point :
The COLA needs to be checked for the square of the window if ultimately you divide by the square of the window [here](https://github.com/scipy/scipy/blob/master/scipy/signal/spectral.py#L1247).

What do you think is better : `check_COLA` takes a window and computes the OLA for the square of it or `check_COLA` remains unchanged and is called for `istft` with a square window as argument?

**Comment by gauss256** [comment]
1. I think `check_COLA` should remain as is. You may want to check with power 1, or 2, or something else altogether depending on your application. The caller of `check_COLA` can have access to all possibilities by keeping the function basic and letting the caller raise the window to the desired power before calling it.

2. I am strongly advocating that `check_COLA` not be called at all from `istft`. If someone wants to check COLA, there is a function for that. If someone wants to use `istft` and doesn't care about COLA, they should be able to do so.

**Comment by e-q** [comment]
It looks like this is closed by #8870. Thanks for putting in the time and effort to set this straight @gauss256!
